### PR TITLE
Issue 1301 simplifications

### LINKED ISSUE #1301: Further expression tree simplifications
Further simplification rules for expression trees:
- broadcast(number) * other_number --> broadcast(number * other_number)
- broadcast(number) * vector --> number * vector
- a/a = ones_like(a)
- a - a = zeros_like(a)

Further simplifications for matrix multiplications:

- associative: A @ (B @ y) = (A @ B) @ y), often A @ B is constant but y is not (e.g. discretized gradient or divergence) and so the whole expression is simplified
- distributive A @ (B + C) = (A @ B + A @ C), in many cases A @ B is constant and then A @ C can be further simplified (e.g. A @ (b + C @ y) + e = (A @ b + e) + (A @ C) @ y, A @ b + e is constant and A @ C is constant
- (A @ c) + (B @ c) should be simplified to (A+B)@c if (A+B) is constant

Automatically do concatenation simplifications

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Adds several automatic simplifications to the expression tree. See #1301 for details.
By informal testing, results in a ~10% speed-up for the models
The `Simplify` function may no longer be needed after this feature is added.

Fixes #1301 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by martinjrobins — CHANGES_REQUESTED** [review]
thanks @tinosulzer , this looks excellent. I had one query below which might be a bug, and I couldn't review the changes to the notebooks, but I wonder if these are just regenerating the same results?

File: pybamm/expression_tree/binary_operators.py [hunk 1]
```diff
@@ -647,6 +654,10 @@ def _binary_evaluate(self, left, right):
         # don't raise RuntimeWarning for NaNs
         return np.minimum(left, right)
 
+    def _binary_new_copy(self, left, right):
+        "See :meth:`pybamm.BinaryOperator._binary_new_copy()`. "
+        return pybamm.minimum(left, right)
```
[the remark this answers is not in the document]
Comment (tinosulzer) [reply pybamm/expression_tree/binary_operators.py:659]: this is under the Minimum class, github is just displaying it compressed

**Comment by tinosulzer** [comment]
@martinjrobins a lot of the missed coverage is now coming from the `Simplify` class not really needing to do anything now. Do you think we should get rid of that class, or keep it?

File: pybamm/expression_tree/binary_operators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pybamm/expression_tree/binary_operators.py:659]: ahhh whoops!

**Comment by martinjrobins** [comment]
> @martinjrobins a lot of the missed coverage is now coming from the `Simplify` class not really needing to do anything now. Do you think we should get rid of that class, or keep it?

just get rid of it if its not doing anything useful. It'll be in the git history if we ever feel the need to revive it

**Review by martinjrobins — APPROVED** [review]

**Pull request merged**